docs/docs/dev_docs/contracts/syntax/functions.md

```diff
@@ -2,53 +2,234 @@
+As alluded to earlier, a private function operates on private information, and is executed by the user. To tell the compiler that this is the kind of function we are creating annotate it with the `#[aztec(private)]` attribute. This will make the [private context](./context.mdx#private-context-broken-down) available within your current function's execution scope.
 
-#include_code functions-SecretFunction /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/main.nr rust
+#include_code redeem_shield /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
 
-## `Public` Functions
+## `unconstrained` functions
 
-To create a public function you can annotate it with the `#[aztec(public)]` attribute. This will make the [public context](../context.mdx#public-context) available within your current function's execution scope.
+Unconstrained functions are are an underlying part of `Noir` and has a deeper explanation [here](https://noir-lang.org/language_concepts/unconstrained). But in short, they are functions which are not directly constrained and therefore should be seem as untrusted! That they are untrusted means that for security the developer must make sure to constrain them when used! 
 
-#include_code functions-OpenFunction /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/main.nr rust
+Beyond using them inside your other functions, they can however be very convenient for providing something that reads storage and apply some logic and returns values to a user interface or similar. Below is a snippet from exposing the `balance_of_private` function from a token implementation, allow a user to easily read their balance. Similar to the `balanceOf` function in the ERC20 standard. 
 
-## `unconstrained` functions
+#include_code balance_of_private /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
 
-#include_code functions-UncontrainedFunction /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/main.nr rust
+:::info
+Note, that unconstrained functions can have access to both public and private data when executed on users device. This is possible since it is not actually part of the circuits that is executed in contract execution.
+:::
 
-## Visibility
+## Oracle functions
+
+An oracle is something that allow us to get data from the outside world into our contracts. The most widely-known types of oracles in blockchain systems are probably Chainlink price feeds, which allow us to get the price of an asset in USD taking non-blockchain data into account. 
+
+While this is one type of oracle, the more general oracle, allows us to get "some" data into the contract. In the context of oracle functions or oracle calls in Aztec, it can essentially be seen as user provided arguments, that can be embedded at any point in the circuit, and thus don't need to be an input parameter. 
```

Comment:
```suggestion
While this is one type of oracle, the more general oracle, allows us to get "some" data into the contract. In the context of oracle functions or oracle calls in Aztec, it can essentially be seen as user-provided arguments, that can be embedded at any point in the circuit, and thus don't need to be an input parameter. 
```